Sources/Wendy/utils/FetchReleases.swift

```diff
@@ -263,39 +272,188 @@ func downloadAsset(_ asset: Release.Asset) async throws -> URL {
+        }
+
+        // Check for path traversal attempts
+        if filePath.contains("../") || filePath.contains("/..") {
+            logger.error(
+                "Archive contains path traversal attempt",
+                metadata: [
+                    "archive": "\(url.path)",
+                    "malicious_path": "\(filePath)",
+                ]
+            )
+            throw ExtractError.maliciousArchive(
+                "Archive contains path traversal: \(filePath)"
+            )
+        }
+
+        // Check for paths that would escape extraction directory when normalized
+        let normalizedPath = (filePath as NSString).standardizingPath
+        if normalizedPath.hasPrefix("/") || normalizedPath.contains("..") {
+            logger.error(
+                "Archive contains normalized path that escapes extraction directory",
+                metadata: [
+                    "archive": "\(url.path)",
+                    "original_path": "\(filePath)",
+                    "normalized_path": "\(normalizedPath)",
+                ]
+            )
+            throw ExtractError.maliciousArchive(
+                "Archive contains escaping path: \(filePath)"
+            )
+        }
+    }
+
+    logger.debug(
+        "Tar archive validation passed",
+        metadata: [
+            "file_count": "\(files.count)"
+        ]
+    )
+
+    // Ensure cleanup on failure
+    var shouldCleanup = true
+    defer {
+        if shouldCleanup {
+            do {
+                try FileManager.default.removeItem(at: extractDir)
+            } catch {
+                let logger = Logger(label: "sh.wendy.utils.fetchReleases")
+                logger.warning(
+                    "Failed to clean up extraction directory on failure",
+                    metadata: [
+                        "path": "\(extractDir.path)",
+                        "error": "\(error)",
+                    ]
+                )
+            }
+        }
+    }
+
+    let extractResult = try await Subprocess.run(
+        Subprocess.Executable.name("tar"),
+        arguments: Subprocess.Arguments(["-xzf", url.path, "-C", extractDir.path]),
+        error: .string(limit: .max)
+    )
+
+    guard extractResult.exitCode == 0 else {
+        let errorMessage = extractResult.standardError ?? "unknown error"
+        logger.error(
+            "Failed to extract tar.gz archive",
+            metadata: [
+                "exit_code": "\(extractResult.exitCode)",
+                "archive": "\(url.path)",
+                "error": "\(errorMessage)",
+            ]
+        )
         throw ExtractError.failedToExtract
     }
 
-    // Find the binary in the extracted directory
-    let enumerator = FileManager.default.enumerator(at: extractDir, includingPropertiesForKeys: nil)
+    // Find the binary in the extracted directory with depth limit
+    // Limit to 5 levels deep to prevent malicious archives from causing excessive traversal
```

Comment:
Shouldn't we also limit the amount of directories scanned in total if we care about this?